[FLINK-12269][table-blink] Support Temporal Table Join in blink planner and runtime

wuchong opened the pull request:
## What is the purpose of the change

Support translate "FOR SYSTEM_TIME AS OF" query into temporal table join for both Batch and Stream.

## Brief change log

  - Introduce some API interface to support lookup table source
    - `LookupableTableSource`, `AsyncTableFunction`, `DefinedPrimaryKey`, `DefinedTableIndex`, `TableIndex`, `LookupConfig`
 - Support for generating optimized logical & physical plan for temporal table join for STREAM and BATCH.
 - Introduce temporal table join operators and async join operators for runtime.
 
Some differences between this PR and blink branch.

 - Support returns multiple lines for async temporal table join
 - Only allow `FOR SYSTEM_TIME AS OF` on left table's proctime field, not a `PROCTIME()` builtin function. This makes syntax clean.
 - Lookup function support external types (i.e. `String`, `Timestamp`...).


## Verifying this change

- Add several validation tests, plan tests , and integration tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

wuchong commented:
Hi @godfreyhe , I add the `JoinPushExpressionsRule.INSTANCE` rule into predicate pushdown stage to make sure the calculation expression can be pushed down (to make sure `ON a+1 = b` is an equi join).  That's why several results of plan test are changed. I have checked, the changes are reasonable.

It would be nice if you can have a look too.

KurtYoung commented on flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/join/temporal/TemporalTableJoinAsyncRunner.java:
```diff
@@ -0,0 +1,268 @@
+/**
+ * The async join runner to lookup the dimension table.
+ */
+public class TemporalTableJoinAsyncRunner extends RichAsyncFunction<BaseRow, BaseRow> {
```
add tests for these runners?

wuchong replied:
Sure

KurtYoung commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/codegen/CalcCodeGenerator.scala:
```diff
@@ -69,6 +71,42 @@ object CalcCodeGenerator {
     new OneInputOperatorWrapper(genOperator)
   }
 
+  private[flink] def generateFunction[T <: Function](
```
why don't adding this to `FunctionCodeGenerator`?

wuchong replied:
Because it actually generate code for a `CalcProgram`, it needs to access the private method `generateProcessCode` in this file. And `FunctionCodeGenerator` only accepts code body as  parameter, not the `CalcProgram.

What do you think about renaming the method name to `generateCalcFunction` to align with `generateCalcOperator` ?

KurtYoung commented on flink-table/flink-table-planner-blink/src/test/resources/org/apache/flink/table/plan/batch/sql/join/TemporalTableJoinTest.xml:
```diff
@@ -0,0 +1,335 @@
+    <Resource name="sql">
+      <![CDATA[
+SELECT * FROM MyTable AS T
+JOIN temporalTest FOR SYSTEM_TIME AS OF T.proctime AS D
```
Why we want to attach the proctime to table `T`, as opposed to just use `proctime()`?

wuchong replied:
Because I think using `proctime()` here is a little semantic ambiguous. Temporal table join is joining the snapshot of temporal table, the version of snapshot is determined by left table's time field. Say `left.proctime`, it means joining the current version of temporal table.

But `temporalTest FOR SYSTEM_TIME AS OF PROCTIME()` is not related to left table, it only returns one (latest) version of temporal table and then be joined with left table. It is the same with 
```sql
SELECT * 
FROM left AS L, 
JOIN (
  SELECT * FROM temporalTest FOR SYSTEM_TIME AS OF PROCTIME()
) AS R ON L.id = R.id
```

That's why I want to restrict it a bit more in the first version. We can discuss it, and can support `PROCTIME()` if we think it is needed. It should be supported easily with several lines change.

KurtYoung replied:
I think you are right. The example you gave will cause different behavior with what we expected. The `protime()` can be translated to `the processing time we begin to scan right table`, and in this case, right table will become a static table. This is not we want.

wuchong commented:
Hi @KurtYoung , I have addressed the review comments.

1. add unit tests for all join runners
2. fix code generation problem when temporal join with udf filter (a bug found recently in internal)

KurtYoung commented on flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/sources/LookupableTableSource.java:
```diff
@@ -0,0 +1,51 @@
+ * @param <T> type of the result
+ */
+@PublicEvolving
+public interface LookupableTableSource<T> extends TableSource<T> {
```
What's the relationship between this `LookupableTableSource` and `DefinedIndexes` and `DefinedPrimaryKey`? Does this table source support getting lookup function for random keys, even if the columns have no index on them or not the primary key?

wuchong replied:
Currently, the lookup key of `LookupableTableSource` must be either primary key or index. Therefore, a concrete `LookupableTableSource` should also implement `DefinedIndexes` or `DefinedPrimaryKey`.

> Does this table source support getting lookup function for random keys, even if the columns have no index on them or not the primary key?

No. It will throw exceptions.

KurtYoung replied:
Do we have any better solution for this? It's hard for user to realize that `LookupableTableSource` should be used with `DefinedIndexes` or `DefinedPrimaryKey`

wuchong replied:
As discussed offline, we will keep `DefinedIndexes` and `DefinedPrimaryKey` in the first version. And integrate them into table source v2 interface in the future.

KurtYoung commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/rules/physical/common/CommonTemporalTableJoinRule.scala:
```diff
@@ -0,0 +1,134 @@
+  override def matches(call: RelOptRuleCall): Boolean = {
+    val join = call.rel[FlinkLogicalJoin](0)
+    val tableScan = call.rel[TableScan](3)
+    matches(join, tableScan)
```
we should also check if it's snapshotted by proctime?

wuchong replied:
Currently, whether it is snapshotted by proctime or rowtime will both be translated into temporal table join. And will throw exception if it is rowtime when physical translation, because we don't support rowtime temporal join currently.

So I think we don't need to check proctime here.

KurtYoung replied:
I noticed current TemporalTableJoin is a `SingleRel`, which is not suitable for further extension after we support scanning data into state and provide event time join. So i think it's inappropriate to translate snapshotted with rowtime to `TemporalTableJoin` and then throw exception inside this physical operator.

wuchong replied:
Make sense. 

I'm also rethinking about the physical node name of `TemporalTableJoin` and `TemporalTableFunctionJoin` (i.e. temporal join in Flink). The node names are really confused to users. Actually, the `TemporalTableJoin` is joining a dimension table which is a SingleRel. 

So how about renaming `TemporalTableJoin` to `DimensionTableJoin`, and renaming `TemporalTableFunctionJoin` to `TemporalJoin`. And we can change the translation rule to:

- snapshot on proctime & table source supports `LookupableTableSource` ==> `DimensionTableJoin` 
- snapshot on rowtime & table source ONLY supports `LookupableTableSource` ==> exception
- snapshot on proctime/rowtime & table source supports scanning ==> `TemporalJoin` 

What do you think ?

KurtYoung replied:
+1 to this idea. How about rename `DimensionTableJoin` to `LookupJoin`, and rename `TemporalJoin` to `TemporalTableJoin`. 

Conceptually `LookupJoin` is very similar with `NestedLoopJoin` but is a `SingleRel`.

wuchong replied:
+1 to `LookupJoin`.

The only concern of `TemporalTableJoin` is that it is actually joining two streams, but there is a `Table` in the name. Do you think it matters?

KurtYoung replied:
Because i think TemporalTable is more consistent with sql standard. `TemporalJoin` will confuse others about what is temporal? It's not intuitive that this is actually used for temporal table join.  You can consider this name as `TemporalTableJoin ' Join`

wuchong replied:
Ok, let's go forward with `LookupJoin` and `TemporalTableJoin`.

wuchong commented:
Hi @KurtYoung , I renamed the node name to `LookupJoin` and move the snapshot on proctime verifying to rules in the second commit: `rename to LookupJoin`.

KurtYoung commented on flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/collector/TableFunctionCollector.java:
```diff
@@ -0,0 +1,89 @@
+
+	@Override
+	public void collect(T record) {
+		this.collected = true;
```
Shouldn't this method call `collector.collect`?

KurtYoung replied:
and i just found that the sub-class of this class `RowToBaseRowCollector` is actually calling `      getCollector.asInstanceOf[Collector[BaseRow]].collect(result)`,

KurtYoung replied:
What i meant is this class already holds a `private Collector<?> collector;`, and also implements `Collector<T>` interface. In this `collect(T record)` method, why don't you collect the record with your own collector, but rely on the caller to first `getCollector()` and collect records?

wuchong replied:
Because `TableFunctionCollector` abstract class does't know the what is the record type `T`. It might be a `BaseRow`, but can also be a `Row` (see `RowToBaseRowCollector`). As a result, `TableFunctionCollector` can't collect by himself, because he doesn't know what the result record is.

KurtYoung replied:
Then i think this class shouldn't be a collector, aka shouldn't implementing `Collector<T>` interface

wuchong replied:
We have to keep the `collect(T record)` interface, because it used to accept UDTF's result value.

What about provide a `outputResult` method used to collect final result. And remove `getCollector()` method and `collect(T record)` implementation. 

```java
	public void outputResult(Object result) {
		this.collected = true;
		collector.collect(result);
	}
```

All the sub-classes can use `outputResult(result)` to collect final result instead of `getCollector().collect(result)`.

KurtYoung replied:
yes, let's try this.

KurtYoung commented on flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/collector/TableFunctionResultFuture.java:
```diff
@@ -0,0 +1,68 @@
+	/**
+	 * Sets the current collector, which used to emit the final row.
+	 */
+	public void setCollector(ResultFuture<?> resultFuture) {
```
setResultFuture?

wuchong replied:
OK

KurtYoung commented on flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/collector/TableFunctionResultFuture.java:
```diff
@@ -0,0 +1,68 @@
+	/**
+	 * Gets the internal collector which used to emit the final row.
+	 */
+	public ResultFuture<?> getCollector() {
```
getResultFuture?

KurtYoung commented on flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/collector/TableFunctionResultFuture.java:
```diff
@@ -0,0 +1,68 @@
+/**
+ * The basic implementation of collector for {@link ResultFuture} in table joining.
+ */
+public abstract class TableFunctionResultFuture<T> extends AbstractRichFunction implements ResultFuture<T> {
```
The responsibility of this class and `TableFunctionCollector` is not clean. It looks like they want to do some encapsulation but actually only contain some get and set methods. And the interface contract also seems not consistent, see the comment i left in `TableFunctionCollector`

wuchong replied:
The responsibility of them are used to generate a collector which will do some filters or projections on both left input row and right table row. That's why we have `setInput`, `getInput` and a `setCollector` to set the real underlying collector.

The only difference between them is `TableFunctionCollector` has an additional `collected` flag used to indicate whether right table is empty. Because UDTF may call `collect(T)` zero or several times, we need a way to know whether it is called zero times, so that we can emit an null row for left join. However, `ResultFuture.complete(Collection<OUT> result)` will be called exactly once, so that we don't need a flag to indicate the "zero call", an empty or null `result` is the "zero call".

JingsongLi commented on flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/join/lookup/AsyncLookupJoinRunner.java:
```diff
@@ -0,0 +1,274 @@
+		TableFunctionResultFuture<BaseRow> resultFuture = generatedResultFuture.newInstance(
+			getRuntimeContext().getUserCodeClassLoader());
+		FunctionUtils.setFunctionRuntimeContext(resultFuture, getRuntimeContext());
+		FunctionUtils.openFunction(resultFuture, new Configuration());
```
Consider saving `parameters` of `open` for use here?

wuchong replied:
Sure

JingsongLi commented on flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/join/lookup/AsyncLookupJoinRunner.java:
```diff
@@ -0,0 +1,274 @@
+		TableFunctionResultFuture<BaseRow> resultFuture = generatedResultFuture.newInstance(
+			getRuntimeContext().getUserCodeClassLoader());
+		FunctionUtils.setFunctionRuntimeContext(resultFuture, getRuntimeContext());
+		FunctionUtils.openFunction(resultFuture, new Configuration());
```
Consider close `resultFuture`?

wuchong replied:
Ok

wuchong commented:
Comments addressed

wuchong commented:
Rebased.

wuchong commented:
@KurtYoung  `TableFunctionCollector` has been refactored.

KurtYoung commented:
+1
